models/ai87-fpndetector.py

```diff
@@ -0,0 +1,544 @@
+                # (n_qualified, n_min_score)
+
+                # Non-Maximum Suppression (NMS)
+
+                # A torch.bool tensor to keep track of which predicted boxes to suppress
+                # True implies suppress, False implies don't suppress
+                suppress = torch.zeros((n_above_min_score), dtype=torch.bool).to(self.device)
+                # (n_qualified)
+
+                # Consider each box in order of decreasing scores
+                for box in range(class_decoded_locs.size(0)):
+                    # If this box is already marked for suppression
+                    if suppress[box]:
+                        continue
+
+                    # Suppress boxes whose overlaps (with this box) are greater than maximum
+                    # overlap
+                    # Find such boxes and update suppress indices
+                    suppress = torch.logical_or(suppress, overlap[box] > max_overlap)
+                    # The max operation retains previously suppressed boxes, like an 'OR' operation
+
+                    # Don't suppress this box, even though it has an overlap of 1 with itself
+                    suppress[box] = False
+
+                # Store only unsuppressed boxes for this class
+                image_boxes.append(class_decoded_locs[~suppress])
+                image_labels.append(
+                    torch.LongTensor((~suppress).sum().item() * [c]).to(self.device))
+                image_scores.append(class_scores[~suppress])
+
+            # If no object in any class is found, store a placeholder for 'background'
+            if len(image_boxes) == 0:
+                image_boxes.append(torch.FloatTensor([[0., 0., 1., 1.]]).to(self.device))
+                image_labels.append(torch.LongTensor([0]).to(self.device))
+                image_scores.append(torch.FloatTensor([0.]).to(self.device))
+
+            # Concatenate into single tensors
+            image_boxes = torch.cat(image_boxes, dim=0)  # (n_objects, 4)
+            image_labels = torch.cat(image_labels, dim=0)  # (n_objects)
+            image_scores = torch.cat(image_scores, dim=0)  # (n_objects)
+            n_objects = image_scores.size(0)
+
+            # Keep only the top k objects
+            if n_objects > top_k:
+                image_scores, sort_ind = image_scores.sort(dim=0, descending=True)
+                image_scores = image_scores[:top_k]  # (top_k)
+                image_boxes = image_boxes[sort_ind][:top_k]  # (top_k, 4)
+                image_labels = image_labels[sort_ind][:top_k]  # (top_k)
+
+            # Append to lists that store predicted boxes and scores for all images
+            all_images_boxes.append(image_boxes)
+            all_images_labels.append(image_labels)
+            all_images_scores.append(image_scores)
+
+        return all_images_boxes, all_images_labels, all_images_scores  # lists of length batch_size
+
+    def are_locations_wide(self):
```

Comment:
Should we modify tinierssd and tinierssd-face model for adding are_locations_wide and are_scores_wide functions with this PR?

Reply:
Train.py checks those functions only if they exist and if not treats all outputs as no-wide. Therefore tinierssd and tinierssd-face models will be treated as expected (as their outputs are non-wide). However, I think adding those would be more readable even if not functionally necessary